File: api/src/main/java/org/example/api/domain/ContractParagraph.java
package org.example.api.domain;

import lombok.*;

import javax.persistence.*;
import java.util.List;

@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@Entity
@Table(name = "contract_paragraph")
public class ContractParagraph {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private long id;

    @Column(name = "seq", nullable = false)
    private int seq;

    @Column(name = "title", nullable = false)
    private String title;

    @Column(name = "contents", nullable = false)
    private String contents;

    @Column(name = "necessary")
    private Character necessary;

    @OneToMany(fetch = FetchType.LAZY, cascade = CascadeType.ALL)
    @JoinTable(name = "contract_paragraph_ref",
            joinColumns = @JoinColumn(name = "parent_id"),
            inverseJoinColumns = @JoinColumn(name = "child_id"))
    private List<ContractParagraph> smallParagraphs;

    @Builder
    public ContractParagraph(long id, int seq, String title, String contents, Character necessary, List<ContractParagraph> smallParagraphs) {
        this.id = id;
        this.seq = seq;
        this.title = title;
        this.contents = contents;
        this.necessary = necessary;
        this.smallParagraphs = smallParagraphs;
    }
}

File: web/src/model/ContractParagraph.ts
export class ContractParagraph {
  id?: number = 0
  title = ''
  contents = ''
  seq = 0
  smallParagraphs: ContractParagraph[] = []

  addParagraphs (paragraph: ContractParagraph) {
    this.smallParagraphs.push(paragraph)
  }

  removeParagraph (index: number) {
    this.smallParagraphs.splice(index, 1)
  }

  static new (old: ContractParagraph): ContractParagraph {
    const contractParagraph = new ContractParagraph()
    contractParagraph.title = old.title
    contractParagraph.contents = old.contents
    contractParagraph.seq = old.seq
    for (const sp of old.smallParagraphs) {
      contractParagraph.addParagraphs(ContractParagraph.new(sp))
    }
    return contractParagraph
  }
}
